### ISSUE TITLE
New lint rule: Exhaustive use of positional parameters

### ISSUE THREAD

**Issue by igoriuz** [dart-lang/linter#2120]
**Describe the rule you'd like to see implemented**
Methods/ Constructors are difficult to read when there's too much (e.g. more than 3 or 4) positional parameters. I would suggest to add a rule which wants them to change into named parameters.

**Examples**
```dart
Car(int wheels, int seats, int kilometres, Color color, int windows);

Car(4, 4, 0, Colors.black, 6); // this is difficult to understand
```

The linter could automatically suggest to make them all named. Maybe with the hint that required parameters should be marked with **@required** annotation.

With the suggestion of the linter, it would become:

```dart
Car({int wheels, int seats, int kilometres, Color color, int windows}); 

Car(wheels: 4, seats: 4, kilometres: 0, color: Colors.black, windows: 6); // this is readable
```

**Comment by bwilkerson** [comment]
I suspect that the issue isn't so much a question of how many parameters a function has as it is a question of whether those parameters have the same type.

Some functions with lots of parameters might actually be fine as long as those parameters are of disjoint types (ignoring, for the moment, the general issue of whether it's good to have functions with large numbers of parameters). On the other hand, even a method with only two parameters can sometimes be better if the parameters are named parameters.

What do you think?

**Comment by igoriuz** [comment]
I haven't considered this before, but i think you're mostly right. Having at least two positional parameters with the same type decreases the readability and clearance of the programmers intent. Additionally you would erase source of errors made by programmers which mistakenly mix up the order of positional parameters.

**Comment by pq** [comment]
See also: #1638
